cmd/login/keypair.go

```diff
@@ -0,0 +1,183 @@
+package login
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"time"
+
+	corev1alpha1 "github.com/giantswarm/apiextensions/v3/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/giantswarm/kubectl-gs/pkg/data/domain/keypair"
+)
+
+const (
+	credentialRetryTimeout    = 250 * time.Millisecond
+	credentialMaxRetryTimeout = 2 * time.Second
+
+	credentialKeyCertCRT = "crt"
+	credentialKeyCertKey = "key"
+	credentialKeyCertCA  = "ca"
+)
+
+func generateKeypairUID() string {
+	hash := sha256.New()
+	_, _ = hash.Write([]byte(time.Now().String()))
+
+	uid := fmt.Sprintf("%x", hash.Sum(nil))
+
+	return uid[:16]
+}
+
+func generateKeypair(cluster, organization, ttl string, groups []string, clusterBasePath string) (*keypair.Keypair, error) {
+	keypairUID := generateKeypairUID()
+	keypairName := fmt.Sprintf("%s-%s", cluster, keypairUID)
+	commonName := fmt.Sprintf("%s.%s.k8s.%s", keypairUID, cluster, clusterBasePath)
+
+	certConfig := &corev1alpha1.CertConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      keypairName,
+			Namespace: metav1.NamespaceDefault,
+			Labels: map[string]string{
+				"cert-operator.giantswarm.io/version": "0.1.0",
+				"giantswarm.io/certificate":           keypairUID,
+				"giantswarm.io/cluster":               cluster,
+				"giantswarm.io/organization":          organization,
+			},
+		},
+		Spec: corev1alpha1.CertConfigSpec{
+			Cert: corev1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    true,
+				ClusterComponent:    keypairUID,
+				ClusterID:           cluster,
+				CommonName:          commonName,
+				DisableRegeneration: true,
+				Organizations:       groups,
+				TTL:                 ttl,
+			},
+			VersionBundle: corev1alpha1.CertConfigSpecVersionBundle{
+				Version: "0.1.0",
+			},
+		},
+	}
+
+	r := &keypair.Keypair{
+		CertConfig: certConfig,
+	}
+
+	return r, nil
+}
+
+// tryToCreateKeypairCredential tries to fetch the keypair credential
+// for a couple of times, until the resource is fetched, or until the timeout is reached.
+func tryToGetKeypairCredential(ctx context.Context, keypairService keypair.Interface, name string) (*corev1.Secret, error) {
+	var secret *corev1.Secret
+	var err error
+
+	retryTimer := time.NewTicker(credentialRetryTimeout)
```

Comment:
I think using the exponential backoff timer we use in operators would simplify this code by a bit.

See this code snippet for example:
https://github.com/giantswarm/aws-operator/blob/49022a4201cda8c6ddf2ebd4bde4c111c4877424/service/internal/ebs/ebs.go#L53-L74